### PR TITLE
mm/iob: add iob_free_queue() interface

### PR DESCRIPTION
**Pull request by anchao**
## Summary

mm/iob: rename the iob_free_queue() to iob_destroy_queue()
mm/iob: add iob_free_queue() interface

make the semantics more precise

## Impact

N/A

## Testing

IOB testing

### CODE REVIEW

File: mm/iob/iob_destroy_queue.c [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  /* Detach the list from the queue head so first for safety (should be safe
+   * anyway).
+   */
+
+  iobq           = qhead->qh_head;
+  qhead->qh_head = NULL;
+
+  /* Remove each I/O buffer chain from the queue */
+
+  while (iobq)
```
Comment (gustavonihei) [inline mm/iob/iob_destroy_queue.c:71]:
```suggestion
  while (iobq != NULL)
```
For pointer validation, being more verbose improves readability.

File: mm/iob/iob_destroy_queue.c [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    {
+      /* Remove the I/O buffer chain from the head of the queue and
+       * discard the queue container.
+       */
+
+      iob = iobq->qe_head;
+      DEBUGASSERT(iob);
```
Comment (gustavonihei) [inline mm/iob/iob_destroy_queue.c:78]:
```suggestion
      DEBUGASSERT(iob != NULL);
```
For pointer validation, being more verbose improves readability.

File: mm/iob/iob_destroy_queue.c (hunk 1 above)
Comment (anchao) [reply mm/iob/iob_destroy_queue.c]: Done

**Comment by antmerlino** [comment]
> 
> ## Summary
> 
> mm/iob: rename the iob_free_queue() to iob_destroy_queue()
> mm/iob: add iob_free_queue() interface
> 
> make the semantics more precise
> ## Impact
> 
> N/A
> ## Testing
> 
> IOB testing

@anchao 

Can you expand on why this change is necessary? In particular, I'd like to understand what the difference between free and destroy is. It'd be good to document these differences more thoroughly in the code too.

**Comment by anchao** [comment]
> Can you expand on why this change is necessary? In particular, I'd like to understand what the difference between free and destroy is. It'd be good to document these differences more thoroughly in the code too.

Hi , @antmerlino 

Thanks for review, 
In the process of TCP / IP stack optimization, I have a requirement to delete an entry from the iob queue list, but the current implementation of IOB does not have this support, look at the API implementation further, I find that the free semantics seems to be more suitable for deleting an entry, and the current free processing is to destroy the entire iob queue. I think it seems that destroy is more suitable for this operations

**Comment by xiaoxiang781216** [comment]
@antmerlino do you have more prefer name for the new API? or merge this patch as it.

**Comment by antmerlino** [comment]
@xiaoxiang781216 @anchao 

I've now spent some more time looking at this and I still think we need to improve the naming.  Firstly, free is the better term to use when we are freeing up a resource to be used again. Destroy makes it sound like the resource doesn't exist after, which is not true. 

What you have now named iob_free_queue() - the implementation of which did not exist before, finds the given IOB in the queue and removes it, freeing both the IOB chain itself, and the qentry. Therefore, I propose you name this function one of the following:

- iob_free_qentry_from_queue() <- Long but very descriptive
- iob_freeone_queue()
- iob_queue_freeone()

I would therefore revert the renaming of the original implementation of iob_free_queue(), as I actually think that name was descriptive of what the function did.

Let me know your thoughts - I'm open to other naming but I don't think this change as it stands is very readable.

**Comment by anchao** [comment]
> @xiaoxiang781216 @anchao
> 
> I've now spent some more time looking at this and I still think we need to improve the naming. Firstly, free is the better term to use when we are freeing up a resource to be used again. Destroy makes it sound like the resource doesn't exist after, which is not true.
> 
> What you have now named iob_free_queue() - the implementation of which did not exist before, finds the given IOB in the queue and removes it, freeing both the IOB chain itself, and the qentry. Therefore, I propose you name this function one of the following:
> 
> * iob_free_qentry_from_queue() <- Long but very descriptive
> * iob_freeone_queue()
> * iob_queue_freeone()
> 
> I would therefore revert the renaming of the original implementation of iob_free_queue(), as I actually think that name was descriptive of what the function did.
> 
> Let me know your thoughts - I'm open to other naming but I don't think this change as it stands is very readable.



> @xiaoxiang781216 @anchao
> 
> I've now spent some more time looking at this and I still think we need to improve the naming. Firstly, free is the better term to use when we are freeing up a resource to be used again. Destroy makes it sound like the resource doesn't exist after, which is not true.
> 
> What you have now named iob_free_queue() - the implementation of which did not exist before, finds the given IOB in the queue and removes it, freeing both the IOB chain itself, and the qentry. Therefore, I propose you name this function one of the following:
> 
> * iob_free_qentry_from_queue() <- Long but very descriptive
> * iob_freeone_queue()
> * iob_queue_freeone()
> 
> I would therefore revert the renaming of the original implementation of iob_free_queue(), as I actually think that name was descriptive of what the function did.
> 
> Let me know your thoughts - I'm open to other naming but I don't think this change as it stands is very readable.

Hi @antmerlino 

Sorry for reply late, thanks for your suggestion, 

the api naming your mentioned seems very detailed and easy to understand, but which out of tune with other iob api naming...
how about change the iob_free_queue() to iob_release_queue() ? 

```
rename: iob_free_queue() -> iob_release_queue()  // enqueue the entire iob/chain entries to freelist
new   : iob_free_queue()                         // enqueue the specified iob/chain entries to freelist
```

**Review by antmerlino — CHANGES_REQUESTED** [review]
@anchao 

"but which out of tune with other iob api naming..."

I don't understand what you mean by this.

---

```rename: iob_free_queue() -> iob_release_queue()  // enqueue the entire iob/chain entries to freelist
new   : iob_free_queue()      
```

That doesn't seem to address the problem at all. We can't just keep using synonym's for the same thing. free, release, destroy, they all mean basically the same thing - with free being by far the most inline with the rest of the code. 

I think we should leave iob_free_queue alone. You are adding new functionality, not new. I don't think you should be renaming an existing function that is perfectly well named already.  The naming problem comes from your new function. It does NOT free a queue. It does NOT destroy a queue. So it should not be named either of those things. What it DOES do, is free, or unlink a single entry from the queue.

I'm not sure I have any better suggestions than I gave before, but I'd say it's a bad idea to rename an existing function who's name already makes sense. And I think it's important that we capture the differences between these functions through their naming. The biggest issue is that one function frees a single entry, where the other frees the whole queue. That "single" vs "all" context needs to be captured in the naming somehow.

Apologies for the long response and push-back, but naming is really important, especially in a generic interface used throughout the OS like this.

**Comment by anchao** [comment]
Hi @antmerlino 

Agree with you, in the new submission, I changed the name of the function to the `iob_free_qentry_from_queue()` you suggested before, please help to review again, Thank you!

File: Documentation/reference/os/iob.rst [hunk 3]
```diff
@@ -164,6 +164,7 @@ Public Function Prototypes
   - :c:func:`iob_remove_queue()`
   - :c:func:`iob_peek_queue()`
   - :c:func:`iob_free_queue()`
+  - :c:func:`iob_free_qentry_from_queue()`
```
Comment (xiaoxiang781216) [inline Documentation/reference/os/iob.rst:167]: @antmerlino and @anchao what about we name the new API iob_free_qentry directly without _from_queue suffix?

Comment (xiaoxiang781216) [inline Documentation/reference/os/iob.rst:167]: @anchao and @antmerlino how about name the new function iob_free_queue_qentry?

Comment (antmerlino) [reply Documentation/reference/os/iob.rst:167]: @xiaoxiang781216 works for me

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**